themes/qgrid/source/js/menu.js

```diff
@@ -15,41 +15,231 @@ function toggleVisibility(e) {
 	}
 }
 
-function filter(e) {
+function searchOnLoad() {
+	const active = document.getElementsByClassName('active-topic')[0];
+	const search = getSearch();
+	if (active && search) {
+		pageTextSearch(search);
+		tagsSearch(search);
+		menuItemsSearch(search);
+	}
+}
+
+function searchOnHashChange() {
+	const search = getSearch();
+	updateLinksInMenuItems(search);
+	pageTextSearch(search);
+	tagsSearch(search);
+	menuItemsSearch(search);
+}
+
+function onSearchChange(e) {
 	e.stopImmediatePropagation();
 	const input = e.target || e.srcElement;
-	const filter = input.value.toLowerCase();
+	const search = input.value.toLowerCase();
+	setSearch(search);
+}
+
+function setSearch(search) {
+	if (search) {
+		location.hash = 'search=' + search;
+	}
+	else {
+		location.hash = '';
+	}
+}
+
+function getSearch() {
+	const contains = new RegExp(/^#search=/);
+	if (contains.test(location.hash)) {
+		return location.hash.substr(8).replace(/%20/g, ' ').toLowerCase();
+	}
+	return '';
+}
+
+function updateLinksInMenuItems(search) {
+	const nav = document.getElementById('nav');
+	const menuItems = nav.getElementsByTagName('li');
+	for (let searchTarget of menuItems) {
+		const title = searchTarget.querySelector('.title');
+		if (title) {
+			let href = title.getAttribute('href');
+			if (href.indexOf('#search=') == -1) {
+				href += '#search=';
+			}
+			else {
+				href = href.substr(0, href.indexOf('=') + 1);
+			}
+			if (search == '') {
```

Comment:
this is strange: it can be that above u can add '#search' and here you can remove, please think how to avoid this